security-controls/sast.py

```diff
@@ -0,0 +1,311 @@
+                tools_list = tools_client if tool["isClientSide"] else tools_server
+                tools_list.append(tool_name)
+        record["tools_full"] = tools_data
+        record["tools_client"] = tools_client
+        record["tools_server"] = tools_server
+        repo = get_repo(repos, (organization, repo_name))
+        repo[SAST_LINK] = CODACY_URL.format(org=organization,repo=repo_name)
+        repo[CODACY] = record
+
+def list_codacy(repos: Repos, codacy_org_data: CodacyOrgData, organization: str):
+    '''Pull the organization's Codacy info and add it to repos and codacy_org_data.'''
+
+    list_codacy_org_data(codacy_org_data, organization)
+    list_codacy_repos(repos, organization)
+
+
+def list_sonar(repos: Repos, org_key: str):
+    '''Pull SonarCloud data for an organization and put it in repos.'''
+    logging.debug("SonarCloud organization %s", org_key)
+
+    url = f"{SONARCLOUD_BASE}/components/search_projects?boostNewProjects=true&ps=50"\
+        "&facets=reliability_rating%2Csecurity_rating%2Csqale_rating%2C"\
+        "coverage%2Cduplicated_lines_density%2Cncloc%2Calert_status%2C"\
+        "languages%2Ctags&f=analysisDate%2CleakPeriodDate&"\
+        f"organization={org_key}"
+
+    auth = HTTPBasicAuth(SONARCLOUD_API_KEY, '')
+    res = requests.get(url, auth=auth, timeout=5)
+    json = res.json()
+    if 'components' not in json:
+        logging.error("SonarCloud - error %s", org_key)
+        return
+    components = json['components']
+    for record in components:
+        project_key = record['key']
+        res = requests.get(f"{SONARCLOUD_BASE}/navigation/component?"\
+            f"component={project_key}", auth=auth, timeout=5)
+        project_json = res.json()
+        if 'alm' not in project_json:
+            logging.error("SonarCloud - error no alm in %s", project_key)
+            continue
+        gh_url = project_json['alm']['url']
+        url_parts = gh_url.split('/')
+        gh_org = url_parts[-2]
+        gh_project = url_parts[-1]
+        logging.debug("SonarCloud - repo %s", gh_project)
+        repo = get_repo(repos, (gh_org, gh_project))
+        repo[SAST_LINK] = SONARCLOUD_URL.format(project_key=project_key)
+        repo[SONAR] = project_json #record
+
+
+def get_data() -> Repos:
+    '''Pull data from SAST tools into Firestore.'''
+    repos = Repos()
+    codacy_org_data = CodacyOrgData()
+
+    # initialize repos list from existing security controls
+    repo_list_from_security_controls(repos)
+
+    # add codacy
+    for org in CODACY_ORGS:
+        list_codacy(repos, codacy_org_data, org)
+
+    # add sonarcloud
+    for org_key in SONAR_ORGS.values():
+        list_sonar(repos, org_key)
+
+    # query github for metadata on all above repos
+    list_github(repos)
+    return repos
+
+def update_sast_values():
+    '''Update security-controls and sast-details in Firestore.'''
+
+    try:
+        logging.info("update_sast_values")
+
+        logging.debug("lengths %s %s %s",
```

Comment:
Why are we debugging the length of Codacy, Sonarcloud and Github token?

Reply:
To show if a secret was omitted or corrupt (without logging the secret itself, of course!). Now that it's working, I'll remove this debug message.